plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/Ec2DiscoveryPluginTests.java

```diff
@@ -48,34 +53,71 @@ private void assertNodeAttributes(Settings settings, String url, String expected
         assertNotNull(e.getCause());
         final String msg = e.getCause().getMessage();
         assertTrue(msg, msg.contains("no protocol: bogus"));
     }
 
     public void testNodeAttributesEmpty() throws Exception {
-        final Path zoneUrl = createTempFile();
-        final IllegalStateException e = expectThrows(
-            IllegalStateException.class,
-            () -> getNodeAttributes(Settings.EMPTY, zoneUrl.toUri().toURL().toString())
-        );
-        assertTrue(e.getMessage(), e.getMessage().contains("no ec2 metadata returned"));
+        try (MetadataServer metadataServer = new MetadataServer("/metadata", exchange -> {
+            exchange.sendResponseHeaders(200, -1);
+            exchange.close();
+        })) {
+            final IllegalStateException e = expectThrows(
+                IllegalStateException.class,
+                () -> getNodeAttributes(Settings.EMPTY, metadataServer.metadataUri(), "")
+            );
+            assertTrue(e.getMessage(), e.getMessage().contains("no ec2 metadata returned"));
+        }
     }
 
     public void testNodeAttributesErrorLenient() throws Exception {
-        final Path dne = createTempDir().resolve("dne");
-        assertNodeAttributes(Settings.EMPTY, dne.toUri().toURL().toString(), null);
+        try (var metadataServer = new MetadataServer("/metadata", exchange -> {
+            exchange.sendResponseHeaders(404, -1);
+            exchange.close();
+        })) {
+            assertNodeAttributes(Settings.EMPTY, metadataServer.metadataUri(), "", null);
+        }
+    }
+
+    public void testNodeAttributesWithToken() throws Exception {
+        try (var metadataServer = new MetadataServer("/metadata", exchange -> {
+            assertEquals("imdsv2-token", exchange.getRequestHeaders().getFirst("X-aws-ec2-metadata-token"));
+            exchange.sendResponseHeaders(200, 0);
+            exchange.getResponseBody().write("us-east-1c".getBytes(StandardCharsets.UTF_8));
+            exchange.close();
+        }, "/latest/api/token", exchange -> {
+            assertEquals("PUT", exchange.getRequestMethod());
+            assertEquals("60", exchange.getRequestHeaders().getFirst("X-aws-ec2-metadata-token-ttl-seconds"));
+            exchange.sendResponseHeaders(200, 0);
+            exchange.getResponseBody().write("imdsv2-token".getBytes(StandardCharsets.UTF_8));
+            exchange.close();
+        })) {
+            assertNodeAttributes(Settings.EMPTY, metadataServer.metadataUri(), metadataServer.tokenUri(), "us-east-1c");
+        }
+    }
+
+    public void testTokenMetadataIsNotAvailable() throws Exception {
+        try (var metadataServer = metadataServerWithoutToken()) {
+            assertNodeAttributes(Settings.EMPTY, metadataServer.metadataUri(), metadataServer.tokenUri(), "us-east-1c");
+        }
+    }
+
+    public void testBogusTokenMetadataUrl() throws Exception {
```

Comment:
Can we add another test where the metadata server closes the connection without responding?

Reply:
Sure